neuralmonkey/runners/runner.py

```diff
@@ -1,50 +1,69 @@
+from typing import List
+import numpy as np
 import tensorflow as tf
 
-from neuralmonkey.learning_utils import feed_dicts
-
-# tests: mypy
-
-class GreedyRunner(object):
-    def __init__(self, decoder, batch_size):
-        self.decoder = decoder
-        self.batch_size = batch_size
-
-    def __call__(self, sess, dataset, coders, extra_fetches=None):
-        batched_dataset = dataset.batch_dataset(self.batch_size)
-        decoded_sentences = []
-        evaluated_fetches = []
-
-        if extra_fetches is None:
-            extra_fetches = []
-
-        loss_with_gt_ins = 0.0
-        loss_with_decoded_ins = 0.0
-        batch_count = 0
-        for batch in batched_dataset:
-            decoded = self.decoder.decoded
-            vocab = self.decoder.vocabulary
-            batch_feed_dict = feed_dicts(batch, coders, train=False)
-            batch_count += 1
-
-            # if is a target sentence, compute also the losses
-            # otherwise, just compute zero
-            if dataset.has_series(self.decoder.data_id):
-                losses = [self.decoder.train_loss,
-                          self.decoder.runtime_loss]
-            else:
-                losses = [tf.zeros([]), tf.zeros([])]
-
-            (loss_with_gt_ins, loss_with_decoded_ins), \
-                    decoded, fetches_batch = \
-                    sess.run((losses, self.decoder.decoded, extra_fetches),
-                             feed_dict=batch_feed_dict)
-            decoded_sentences_batch = \
-                    vocab.vectors_to_sentences(decoded)
-
-            decoded_sentences += decoded_sentences_batch
-            evaluated_fetches += [fetches_batch]
-
-        return decoded_sentences, \
-               loss_with_gt_ins / batch_count, \
-               loss_with_decoded_ins / batch_count, \
-               evaluated_fetches
+from neuralmonkey.tf_manager import RunResult
+from neuralmonkey.runners.base_runner import BaseRunner, \
+    Executable, ExecutionResult, NextExecute
+
+# tests: mypy,pylint
+# pylint: disable=too-few-public-methods
+
+
+class GreedyRunner(BaseRunner):
+
+    def __init__(self, output_series, decoder):
+        super(GreedyRunner, self).__init__(output_series, decoder)
+
+    def get_executable(self, train=False, summaries=True):
+        if train:
+            losses = [self._decoder.train_loss,
+                      self._decoder.runtime_loss]
+        else:
+            losses = [tf.zeros([]), tf.zeros([])]
+        to_run = losses + self._decoder.runtime_logprobs
```

Comment:
to počítání nul je blbost, stejně jako pak to počítání těch indexů do toho pole výsledků.
Jako ``fetches`` může ``session.run`` dostat libovolně strukturovanou pythoní věc, takže i dict. Stačí to udělat takhle:

```
fetches = {"logprobs": self._decoder.runtime_logprobs}

if train:
    fethes["losses"] = {"train": self._decoder.train_loss,
                        "runtime": self._decoder.runtime_loss}

return GreedyRunExecutable(self.all_coders, fetches, self._decoder.vocabulary)
```

Jestli teda ``fetches`` nikde někdo nijak jinak nepoužívá, než že to cpe do ``session.run``